### DISCUSSION TITLE
The CodeQL database is not compatible with the QL library

### DISCUSSION THREAD

**Discussion by vanhauser-thc** [github/securitylab#220] [Q&A] [answered]
I think this is a very basic answer, however I find no documentation to solve this, so here we go:

```
cd ~
mkdir semmle ; cd semmle
wget https://github.com/github/codeql-cli-binaries/releases/latest/download/codeql-linux64.zip
unzip codeql-linux64.zip
git clone https://github.com/Semmle/ql
```

Next for a simple example project I created a database with `codeql database create ql.db -l cpp -c make` which worked fine.

Now I want to search for the various security issues that are in the example project. however I whatever I do, "database analyze" or "query run" - I get th efatal error that the ql database and ql library are not compatbible.

```
$  /home/test/semmle/codeql/codeql database analyze ql.db /home/test/semmle/ql/cpp/ql/src/Security/ --format=csv --output=out/out.csv
Picked up _JAVA_OPTIONS: -Dawt.useSystemAAFontSettings=on -Dswing.aatext=true
Running queries.
Picked up _JAVA_OPTIONS: -Dawt.useSystemAAFontSettings=on -Dswing.aatext=true
Compiling query plan for /home/testt/semmle/ql/cpp/ql/src/Security/CWE/CWE-732/DoNotCreateWorldWritable.ql.
Compiling upgrade for /home/test/semmle/ql/cpp/ql/src/Security/CWE/CWE-732/DoNotCreateWorldWritable.ql
A fatal error occurred: The CodeQL database at ql.db
is not compatible with the QL library /home/test/semmle/ql/cpp/ql/src/Security/CWE/CWE-732/DoNotCreateWorldWritable.ql uses.
The database may be too new for the QL libraries the query is using; try upgrading them.
Alternatively, running 'codeql database upgrade ql.db' with an appropriate --search-path option might help.
```

unsurprisingly `codeql database upgrade ql.db` tells me there is nothing to update.

So - how do I update the ql libraries so I can use them?

**Comment by xcorail** [comment]
Hey @vanhauser-thc 👋🏾 
Try to update your CodeQL repo. Don't forget to also update the submodules when you do: `git submodule update --remote`

**Reply by vanhauser-thc** [reply]
Hi, as you can see from my first lines it is freshly cloned.
and this sadly does not change anything:
```
$ cd ~/semmle/ql
$ git submodule init
$ git submodule update --remote
$ git pull
Current branch master is up to date.
```

**Reply by vanhauser-thc** [reply]
if you mean codeql itself, not the ql library - no git, I download the binary release (as can be seen in the how I set things up)

**Comment by vanhauser-thc** [comment, marked as answer]
I removed every codeql and ql directory, redownloaded from scratch, ensured the PATH is fine - and now it works. seems that the ql intelligence to find its libraries is too enthusiastic and goes further and beyond even if --search-path= is specified.

thank you though!
